### PR TITLE
[WIP] Provider GitHub

### LINKED ISSUE #101: [v2] Github external provider

### PR DESCRIPTION
**Pull request by elitan**
close #101

### CODE REVIEW

File: src/routes/auth/providers/github.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      authorizationURL: AUTH_GITHUB_AUTHORIZATION_URL as 'string',
+      tokenURL: AUTH_GITHUB_TOKEN_URL as 'string',
+      userProfileURL: AUTH_GITHUB_USER_PROFILE_URL as 'string',
+      scope: ['user:email']
+    },
+    async (
+      accessToken: string,
+      refreshToken: string,
+      profile: Profile,
+      done: (error: any, user?: any) => void
```
Comment (elitan) [inline src/routes/auth/providers/github.ts:38]: @plmercereau How could I handle this better?

I am using this: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/passport-github2/index.d.ts

Comment (plmercereau) [reply src/routes/auth/providers/github.ts]: @elitan I just push an improvement, I think we can do better, working on it

Comment (plmercereau) [reply src/routes/auth/providers/github.ts]: Nah, even the [typings tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/passport-github2/passport-github2-tests.ts) are not better... Let's park an additional improvement, it's already ok I think

Comment (elitan) [reply src/routes/auth/providers/github.ts]: Got it. Thanks for taking a look.

**Comment by elitan** [comment]
This PR works now. However, we might want to move around some code to make it simple to support Google, Facebook, and Twitter.

Everything in the ` GitHubStrategy` async function and in the `/callback` function is generic and would work with Google, Facebook, Twitter because PassportJS standardized the response from these services.

So something like 

`manageProviderStrategy(_, __, profile: Profile, done: oauth2.VerifyCallback, provider: string): void`

where `provider` is one of `[github, facebook, google, twitter]`.

And for `/callback`:

`manageCallback(req: Request, res: Response)`.

Should we place these two functions in `@shared/provider-helpers.ts`?

**Comment by plmercereau** [comment]
> This PR works now. However, we might want to move around some code to make it simple to support Google, Facebook, and Twitter.
> 
> Everything in the ` GitHubStrategy` async function and in the `/callback` function is generic and would work with Google, Facebook, Twitter because PassportJS standardized the response from these services.
> 
> So something like
> 
> `manageProviderStrategy(_, __, profile: Profile, done: oauth2.VerifyCallback, provider: string): void`
> 
> where `provider` is one of `[github, facebook, google, twitter]`.
> 
> And for `/callback`:
> 
> `manageCallback(req: Request, res: Response)`.
> 
> Should we place these two functions in `@shared/provider-helpers.ts`?

Sure, we'll need to restructure it when we'll code a second Oauth provider to use the same bits

**Comment by elitan** [comment]
`/routes/auth/providers/utils.ts` sounds OK, but will break our standard of only having endpoints under `/routes`.

`AUTH_GITHUB_CALLBACK_URL`:
If we know beforehand what the URI and domain are, then sure. But I'm not sure how we would interfere this value.

On Nhost we autogenerate these URLs.

**Comment by plmercereau** [comment]
> `/routes/auth/providers/utils.ts` sounds OK, but will break our standard of only having endpoints under `/routes`.
> 
> `AUTH_GITHUB_CALLBACK_URL`:
> If we know beforehand what the URI and domain are, then sure. But I'm not sure how we would interfere this that.
> 
> On Nhost we autogenerate these URLs.
> 

AUTH_GITHUB_CALLBACK_URL: we can generate from SERVER_URL that is already defined (only used for emails until now)

**Comment by elitan** [comment]
The last thing we want to do is:

If one of `AUTH_GITHUB_AUTHORIZATION_URL, AUTH_GITHUB_TOKEN_URL, AUTH_GITHUB_USER_PROFILE_URL` is set. All must be set and we need to use :

```
        clientID: AUTH_GITHUB_CLIENT_ID,
        clientSecret: AUTH_GITHUB_CLIENT_SECRET,
        callbackURL: AUTH_GITHUB_CALLBACK_URL,
        authorizationURL: AUTH_GITHUB_AUTHORIZATION_URL,
        tokenURL: AUTH_GITHUB_TOKEN_URL,
        userProfileURL: AUTH_GITHUB_USER_PROFILE_URL,
        scope: ['user:email'],
        passReqToCallback: true
```

Else we should use:

```
        clientID: AUTH_GITHUB_CLIENT_ID,
        clientSecret: AUTH_GITHUB_CLIENT_SECRET,
        callbackURL: AUTH_GITHUB_CALLBACK_URL,
        scope: ['user:email'],
        passReqToCallback: true
```

I don't know how to effectively do that in TS.

**Comment by plmercereau** [comment]
> The last thing we want to do is:
> 
> If one of `AUTH_GITHUB_AUTHORIZATION_URL, AUTH_GITHUB_TOKEN_URL, AUTH_GITHUB_USER_PROFILE_URL` is set. All must be set and we need to use :
> 
> ```
>         clientID: AUTH_GITHUB_CLIENT_ID,
>         clientSecret: AUTH_GITHUB_CLIENT_SECRET,
>         callbackURL: AUTH_GITHUB_CALLBACK_URL,
>         authorizationURL: AUTH_GITHUB_AUTHORIZATION_URL,
>         tokenURL: AUTH_GITHUB_TOKEN_URL,
>         userProfileURL: AUTH_GITHUB_USER_PROFILE_URL,
>         scope: ['user:email'],
>         passReqToCallback: true
> ```
> 
> Else we should use:
> 
> ```
>         clientID: AUTH_GITHUB_CLIENT_ID,
>         clientSecret: AUTH_GITHUB_CLIENT_SECRET,
>         callbackURL: AUTH_GITHUB_CALLBACK_URL,
>         scope: ['user:email'],
>         passReqToCallback: true
> ```
> 
> I don't know how to effectively do that in TS.

Good point. Just pushed the fix

**Pull request merged**